#145028641 Front tests

georgreen opened the pull request:
#### What does this PR do?
- Adds unit tests for the app hc/front
#### Description of Task to be completed?
- Add tests to test team access functionality in channels
- Add tests to check bad kinds of channels do not work
- Add tests to check team access works on checks
- Add tests to check pushover validate priority rejecting invalid ones
- Refactor code base 

#### How should this be manually tested?
- python manage.py test hc/front

#### What are the relevant pivotal tracker stories?
- #145028641

Commit:
[#145028641] FT: Add Test for Team Access
- Tested if team members can access each other channels and profile

davidmukiibi committed:
ft: Added test for bad kinds [#145028641]

- We added a test for testing that only kinds specified work!

Commit:
Chore: Merge  'origin/ft-testing-bad-kinds-#145028641' into ft-Front-tests-145028641

Commit:
[#145028641] Chore: Merged branch badkinds and Channels access
- Merged Branches to have one that present story on pivotal tracker
- All changes on this story will be merge to this branch hence forth

Commit:
Chore: Fixed typo in tests

Commit:
[#145028641] Implemented team access test in Checks
- Added test for team access in checks feature

Commit:
[#145028641] Implemented test for priority validation.
- Added a test for validating priority in pushover feature.

Commit:
[#145028641] Fixed travis build error
- Syntax Error caused by typo in add pushover

collinmutembei commented on hc/front/views.py:
```diff
@@ -330,6 +331,7 @@ def channel_checks(request, code):
         "channel": channel
     }
 
+    print("returned page")
```
I believe this print statement was introduced for debugging purposes. In which case you should avoid pushing code with debug statements.

georgreen replied:
done

collinmutembei reviewed (changes requested):
👀 Still looking into this.

collinmutembei commented:
Remove the square brackets on the PR title.

Commit:
[#145028641] Chore: Removed Debug code
- Removed unnessecary debug code from views.py

collinmutembei commented on hc/front/tests/test_add_channel.py:
```diff
@@ -37,5 +36,30 @@ def test_instructions_work(self):
+
+            url = "/integrations/{}/checks/".format(self.channel.code)
+
+            self.client.login(username="bob@example.org", password="password")
```
Do lines [51-53](https://github.com/andela/healthchecks-thunknalla/blob/a066e49909c4a8b039d53bebf7047853310afff4/hc/front/tests/test_add_channel.py#L51-L53) need to be in the for loop?

georgreen replied:
After looking at it, I think logging in once is enough and we don't need line 51 in the loop, we will move it out of the loop.

collinmutembei commented on hc/front/tests/test_add_channel.py:
```diff
@@ -37,5 +36,30 @@ def test_instructions_work(self):
+        bad_kinds = ('0798885255', 'twitter', 'instagram', 'whatsapp',
+                     'telegram', 'facebook', 'reddit')
+        for any_kind in bad_kinds:
+            url = "/integrations/add_%s/" % any_kind
```
I recommend using `str.format()` for string formatting.

Resource: [pyformat.info](https://pyformat.info)

collinmutembei commented on hc/front/tests/test_add_channel.py:
```diff
@@ -37,5 +36,30 @@ def test_instructions_work(self):
+        for any_kind in bad_kinds:
+            url = "/integrations/add_%s/" % any_kind
+            r = self.client.get(url)
+            assert r.status_code == 404
```
Replace this assert statement with its equivalent in Django.

Resource: [Django assertions](https://docs.djangoproject.com/en/1.10/topics/testing/tools/#assertions)

collinmutembei commented on hc/front/tests/test_add_check.py:
```diff
@@ -12,3 +12,18 @@ def test_it_works(self):
+        self.assertRedirects(r, "/checks/")
+
+        check = Check.objects.get(code=self.check.code)
+        assert check.timeout.total_seconds() == 5000
```
Use djangoish assert statements. See previous similar comment.

collinmutembei commented on hc/front/tests/test_add_check.py:
```diff
@@ -12,3 +12,18 @@ def test_it_works(self):
+        url = "/checks/%s/timeout/" % self.check.code
+        payload = {"timeout": 5000, "grace": 60}
+
+        r = self.client.post(url, data=payload)
```
Refrain from using variable names such as `r`. Use a descriptive name. If you find such in existing code, try to understand the context and update them to a meaningful name.

@georgreen @davidmukiibi @victoriaaoka @Andretalik @Kachulio1

collinmutembei commented on hc/front/tests/test_add_pushbover.py:
```diff
@@ -38,3 +39,15 @@ def test_it_validates_nonce(self):
         assert r.status_code == 403
 
     ### Test that pushover validates priority
+    def test_validate_priority(self):
```
@georgreen @davidmukiibi Could either of you explain what is going on in this test method.

georgreen replied:
@collinmutembei pushover is a service that provides a simple way to send notifications to devices from our web app, the notification sent have priority ranging from -2 to 2, this function test that we can only add pushover notification in this valid range so anything else should not be validated ie strings and integers out of range. If the user tries to add such a pushover without the correct priority the response status code should be 400 (bad request). Hope am clear enough we can expand further on our reasoning in our stand up meeting.

collinmutembei replied:
🆒

collinmutembei reviewed (changes requested):
@georgreen @davidmukiibi Have a look at the comments I have left and attend to them.

Commit:
[##145028641] Ft: Fixed repetition in Tests
- Improved test design by removing unnessecary code from the for loop

Commit:
[#145028641] Chore: Refactored to improve readability
- Gave variables more descriptive names to improve readability

Commit:
[#145028641] Ft: Implemented more tests for Pushover
- Added more test to test the pushover functionality
- Test added covered a wider range includiong special cases and previous unseen scenarios

Commit:
[#145028641] Chore: Refactored code for readability
- Improved the variable naming style to be more descriptive
- Pep 8 rules where also apllied where nessecary

collinmutembei commented:
Looks good. Add the PR description and I'll merge.
